[CI] QueryPhaseTimeoutTests testScorerTimeoutPoints failing

**Build Scans:**
- [elasticsearch-intake #19479 / part1](https://gradle-enterprise.elastic.co/s/b233quyczq3am)
- [elasticsearch-periodic-platform-support #6780 / opensuse-leap-15_platform-support-unix](https://gradle-enterprise.elastic.co/s/4hu2zg7gk4qmm)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.search.query.QueryPhaseTimeoutTests.testScorerTimeoutPoints" -Dtests.seed=FC34D732AE5ADC5A -Dtests.locale=und -Dtests.timezone=Europe/Madrid -Druntime.java=23
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.search.query.QueryPhaseTimeoutTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testScorerTimeoutPoints),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected:<2048> but was:<2399>
```

**Issue Reasons:**
- [main] 2 failures in test testScorerTimeoutPoints (0.9% fail rate in 213 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/server/src/test/java/org/elasticsearch/search/query/QueryPhaseTimeoutTests.java
+++ b/server/src/test/java/org/elasticsearch/search/query/QueryPhaseTimeoutTests.java
@@ -116,6 +116,7 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
     }
 
     private static ContextIndexSearcher newContextSearcher(IndexReader reader) throws IOException {
+        // note that no executor is provided, as this test requires sequential execution
         return new ContextIndexSearcher(
             reader,
             IndexSearcher.getDefaultSimilarity(),
@@ -125,27 +126,35 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
         );
     }
 
-    public void testScorerTimeoutTerms() throws IOException {
+    /**
+     * Test that a timeout is appropriately handled when the (exitable) directory reader raises it while loading terms enum
+     * as the scorer supplier is requested.
+     */
+    public void testScorerSupplierTimeoutTerms() throws IOException {
         assumeTrue("Test requires more than one segment", reader.leaves().size() > 1);
         int size = randomBoolean() ? 0 : randomIntBetween(100, 500);
-        scorerTimeoutTest(size, context -> {
+        scorerSupplierTimeoutTest(size, context -> {
             final TermsEnum termsEnum = context.reader().terms("field").iterator();
             termsEnum.next();
         });
     }
 
-    public void testScorerTimeoutPoints() throws IOException {
+    /**
+     * Test that a timeout is appropriately handled when the (exitable) directory reader raises it while loading points
+     * as the scorer supplier is requested.
+     */
+    public void testScorerSupplierTimeoutPoints() throws IOException {
         assumeTrue("Test requires more than one segment", reader.leaves().size() > 1);
         int size = randomBoolean() ? 0 : randomIntBetween(100, 500);
-        scorerTimeoutTest(size, context -> {
+        scorerSupplierTimeoutTest(size, context -> {
             PointValues pointValues = context.reader().getPointValues("long");
             pointValues.size();
         });
     }
 
-    private void scorerTimeoutTest(int size, CheckedConsumer<LeafReaderContext, IOException> timeoutTrigger) throws IOException {
+    private void scorerSupplierTimeoutTest(int size, CheckedConsumer<LeafReaderContext, IOException> timeoutTrigger) throws IOException {
         {
-            TimeoutQuery query = newMatchAllScorerTimeoutQuery(timeoutTrigger, false);
+            TimeoutQuery query = newMatchAllScorerSupplierTimeoutQuery(timeoutTrigger, false);
             try (SearchContext context = createSearchContext(query, size)) {
                 QueryPhase.executeQuery(context);
                 assertFalse(context.queryResult().searchTimedOut());
@@ -154,18 +163,20 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
             }
         }
         {
-            TimeoutQuery query = newMatchAllScorerTimeoutQuery(timeoutTrigger, true);
+            TimeoutQuery query = newMatchAllScorerSupplierTimeoutQuery(timeoutTrigger, true);
             try (SearchContext context = createSearchContextWithTimeout(query, size)) {
                 QueryPhase.executeQuery(context);
                 assertTrue(context.queryResult().searchTimedOut());
                 int firstSegmentMaxDoc = reader.leaves().get(0).reader().maxDoc();
-                assertEquals(Math.min(2048, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.totalHits.value());
+                // we are artificially raising the timeout when pulling the scorer supplier.
+                // We score the entire first segment, then trigger timeout.
+                assertEquals(firstSegmentMaxDoc, context.queryResult().topDocs().topDocs.totalHits.value());
                 assertEquals(Math.min(size, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.scoreDocs.length);
             }
         }
     }
 
-    private static TimeoutQuery newMatchAllScorerTimeoutQuery(
+    private static TimeoutQuery newMatchAllScorerSupplierTimeoutQuery(
         CheckedConsumer<LeafReaderContext, IOException> timeoutTrigger,
         boolean isTimeoutExpected
     ) {
@@ -177,6 +188,7 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
 
                     @Override
                     public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                        // trigger the timeout as soon as the scorer supplier is request for the second segment
                         if (firstSegment == false && isTimeoutExpected) {
                             shouldTimeout = true;
                         }
@@ -190,6 +202,96 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
         };
     }
 
+    /**
+     * Test that a timeout is appropriately handled when the (exitable) directory reader raises it while loading terms enum
+     * as the scorer is retrieved from the scorer supplier
+     */
+    public void testScorerGetTimeoutTerms() throws IOException {
+        assumeTrue("Test requires more than one segment", reader.leaves().size() > 1);
+        int size = randomBoolean() ? 0 : randomIntBetween(100, 500);
+        scorerGetTimeoutTest(size, context -> {
+            final TermsEnum termsEnum = context.reader().terms("field").iterator();
+            termsEnum.next();
+        });
+    }
+
+    /**
+     * Test that a timeout is appropriately handled when the (exitable) directory reader raises it while loading points
+     * as the scorer is retrieved from the scorer supplier
+     */
+    public void testScorerGetTimeoutPoints() throws IOException {
+        assumeTrue("Test requires more than one segment", reader.leaves().size() > 1);
+        int size = randomBoolean() ? 0 : randomIntBetween(100, 500);
+        scorerGetTimeoutTest(size, context -> {
+            PointValues pointValues = context.reader().getPointValues("long");
+            pointValues.size();
+        });
+    }
+
+    private void scorerGetTimeoutTest(int size, CheckedConsumer<LeafReaderContext, IOException> timeoutTrigger) throws IOException {
+        {
+            TimeoutQuery query = newMatchAllScorerGetTimeoutQuery(timeoutTrigger, false);
+            try (SearchContext context = createSearchContext(query, size)) {
+                QueryPhase.executeQuery(context);
+                assertFalse(context.queryResult().searchTimedOut());
+                assertEquals(numDocs, context.queryResult().topDocs().topDocs.totalHits.value());
+                assertEquals(size, context.queryResult().topDocs().topDocs.scoreDocs.length);
+            }
+        }
+        {
+            TimeoutQuery query = newMatchAllScorerGetTimeoutQuery(timeoutTrigger, true);
+            try (SearchContext context = createSearchContextWithTimeout(query, size)) {
+                QueryPhase.executeQuery(context);
+                assertTrue(context.queryResult().searchTimedOut());
+                int firstSegmentMaxDoc = reader.leaves().get(0).reader().maxDoc();
+                // we are artificially raising the timeout when pulling the scorer supplier.
+                // We score the entire first segment, then trigger timeout.
+                assertEquals(firstSegmentMaxDoc, context.queryResult().topDocs().topDocs.totalHits.value());
+                assertEquals(Math.min(size, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.scoreDocs.length);
+            }
+        }
+    }
+
+    private static TimeoutQuery newMatchAllScorerGetTimeoutQuery(
+        CheckedConsumer<LeafReaderContext, IOException> timeoutTrigger,
+        boolean isTimeoutExpected
+    ) {
+        return new TimeoutQuery() {
+            @Override
+            public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+                return new MatchAllWeight(this, boost, scoreMode) {
+                    boolean firstSegment = true;
+
+                    @Override
+                    public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                        ScorerSupplier scorerSupplier = super.scorerSupplier(context);
+                        return new ScorerSupplier() {
+                            @Override
+                            public Scorer get(long leadCost) throws IOException {
+                                // trigger the timeout as soon as the scorer is requested for the second segment
+                                if (firstSegment == false && isTimeoutExpected) {
+                                    shouldTimeout = true;
+                                }
+                                timeoutTrigger.accept(context);
+                                assert shouldTimeout == false : "should have already timed out";
+                                firstSegment = false;
+                                return scorerSupplier.get(leadCost);
+                            }
+
+                            @Override
+                            public long cost() {
+                                return scorerSupplier.cost();
+                            }
+                        };
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Test that a timeout is appropriately handled while bulk scoring, via cancellable bulk scorer
+     */
     public void testBulkScorerTimeout() throws IOException {
         int size = randomBoolean() ? 0 : randomIntBetween(100, 500);
         {
@@ -207,6 +309,8 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
                 QueryPhase.executeQuery(context);
                 assertTrue(context.queryResult().searchTimedOut());
                 int firstSegmentMaxDoc = reader.leaves().get(0).reader().maxDoc();
+                // See CancellableBulkScorer#INITIAL_INTERVAL for the source of 2048: we always score the first
+                // batch of up to 2048 docs, and only then raise the timeout
                 assertEquals(Math.min(2048, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.totalHits.value());
                 assertEquals(Math.min(size, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.scoreDocs.length);
             }
@@ -233,7 +337,7 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
                             }
 
                             @Override
-                            public BulkScorer bulkScorer() throws IOException {
+                            public BulkScorer bulkScorer() {
                                 final float score = score();
                                 final int maxDoc = context.reader().maxDoc();
                                 return new BulkScorer() {
@@ -251,7 +355,7 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
                                         }
                                         if (timeoutExpected) {
                                             // timeout after collecting the first batch of documents from the 1st segment, or the entire 1st
-                                            // segment
+                                            // segment if max > firstSegment.maxDoc()
                                             shouldTimeout = true;
                                         }
                                         return max == maxDoc ? DocIdSetIterator.NO_MORE_DOCS : max;
@@ -274,6 +378,9 @@ public class QueryPhaseTimeoutTests extends IndexShardTestCase {
         TestSearchContext context = new TestSearchContext(null, indexShard, newContextSearcher(reader)) {
             @Override
             public long getRelativeTimeInMillis() {
+                // this controls whether a timeout is raised or not. We abstract time away by pretending that the clock stops
+                // when a timeout is not expected. The tiniest increment to relative time in millis triggers a timeout.
+                // See QueryPhase#getTimeoutCheck
                 return query.shouldTimeout ? 1L : 0L;
             }
         };

